guide_service_errors.md

```diff
@@ -0,0 +1,234 @@
+to guarantee which errors a client should expect to encounter.
+
+## Step 1: Define your Service Errors
+
+First, you need to determine which service error codes your service may return, and what
+information will be contained in error responses corresponding with these codes. You'll
+need to define four things:
+
+- The application-specific string code associated with some service error
+- The HTTP status it'll return
+- The string message it'll return
+- The schema defining the format of the error details
+
+For example, say we want to standardize the way errors are returned as a result of violating
+the service's rate-limiting constraint. We would need to pick a string error code such as
+`QUOTA_EXCEEDED`, an HTTP status such as `429`, a string message describing the failure, and
+some schema describing the custom error details sent back to the client.
+
+## Step 2: Write Your Service Error Definition
+
+Write a Java enum implementing the `ServiceError` interface, which will contain a set of
+service error definitions for your service. The enum implementation may expose methods
+for obtaining the HTTP status, the error code, the error message, and the error detail type.
+Here is an example implementation which defines one service error and exposes all four
+accessor methods:
+
+```java
+public enum SampleServiceError implements ServiceError
+{
+  QUOTA_EXCEEDED(HttpStatus.S_429_TOO_MANY_REQUESTS, "You've exceeded your daily request quota.", QuotaDetails.class),
+  INVALID_PERMISSIONS(HttpStatus.S_403_FORBIDDEN, "User does not have valid permissions", null),
+  ILLEGAL_PARAM(HttpStatus.S_400_BAD_REQUEST, "Request parameter cannot be used", null);
+  // Add more service errors here...
+
+  SampleServiceError(HttpStatus status, String message, Class<? extends RecordTemplate> errorDetailType)
+  {
+    _httpStatus = status;
+    _message = message;
+    _errorDetailType = errorDetailType;
+  }
+
+  private final HttpStatus _status;
+  private final String _message;
+  private final Class<? extends RecordTemplate> _errorDetailType;
+
+  /*
+   * Suggested: Define string constants for each of the error codes for easy access.
+   * These codes should match the enum value names.
+   */
+  public class Codes
+  {
+    QUOTA_EXCEEDED = "QUOTA_EXCEEDED";
+    INVALID_PERMISSIONS = "INVALID_PERMISSIONS";
+  }
+
+  @Override
+  public HttpStatus httpStatus()
+  {
+    return _status;
+  }
+
+  @Override
+  public String code()
+  {
+    return name();
+  }
+
+  @Override
+  public String message()
+  {
+    return _message;
+  }
+
+  @Override
+  public Class<? extends RecordTemplate> errorDetailType()
+  {
+    return _errorDetailType;
+  }
+}
+```
+
+## Step 3: Apply the Service Errors to your Resource
+
+Now that you have defined your service errors, you can use them to specify which resources
+or particular resource methods may return such errors. You can do this by using a few
```

Comment:
Can we say that in java, you can do this by using annotations as described below?

Reply:
Not sure what you mean by this. Do you think it would be better if I renamed this page to "Configuring Service Errors in Java"?

Reply:
Oh, it's related to my previous comment related to `java`. My comment was to suggest to replace `You can do this by using a few...` by `In java, you can do this by using annotations as described below`